core/chains/evm/client/tx_simulator.go

```diff
@@ -0,0 +1,134 @@
+package client
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink/v2/common/config"
+
+	commonclient "github.com/smartcontractkit/chainlink/v2/common/client"
+)
+
+const ErrOutOfCounters = "not enough counters to continue the execution"
+
+type simulatorClient interface {
+	CallContext(ctx context.Context, result interface{}, method string, args ...interface{}) error
+}
+
+// ZK chains can return an out-of-counters error
+// This method allows a caller to determine if a tx would fail due to OOC error by simulating the transaction
+// Used as an entry point for custom simulation across different chains
+func SimulateTransaction(ctx context.Context, client simulatorClient, lggr logger.SugaredLogger, chainType config.ChainType, msg ethereum.CallMsg) error {
+	var err error
+	switch chainType {
+	case config.ChainZkEvm:
+		err = simulateTransactionZkEvm(ctx, client, lggr, msg)
+	default:
+		err = simulateTransactionDefault(ctx, client, msg)
+	}
+	// ClassifySendError will not have the proper fields for logging within the method due to the empty Transaction passed
+	code := ClassifySendError(err, lggr, &types.Transaction{}, msg.From, chainType.IsL2())
```

Comment:
I think it was a suggestion from Dimitris so we could take advantage of configurable errors after Nick's changes unless I misunderstood.  You bring up a good point though that this would be mixing different types of errors though. I could setup a different parser for tx simulation errors but would that be overkill?